### PR TITLE
Simplify SHA gadgets and add a stdlib

### PR DESCRIPTION
**Pull request by stefandeml**
With this PR  we refactor the integration of optimised gadgets for SHA2 computations.

Goals:
*reduce dependency on the backend
*push most most of the user logic to the ZoKrates DSL

Changes:
*reduce to the number libsnark gadgets to one: LIBSNARK/SHAROUND. This gadgets does one round of SHA256 with provided initialisation vector.
*add a standard library located at `./stdlib/` in the ZoKrates root folder. All other logic is described here using the ZoKrates DSL.

thanks to @sanchopansa and @Guthl for the support

Open topics:
- [x] Decide on where to apply bool checks. E.g. `sha256packed`
- [x] Add tests for the stdlib

### CODE REVIEW

File: stdlib/hashes/sha256/512bitPadded.code [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import "./1024bit.code" as sha256 
+
+// A function that takes 2 field[256] arrays as inputs 
+// and returns their sha256 compression function as an array of 256 field elements.
```
Comment (Schaeff) [inline stdlib/hashes/sha256/512bitPadded.code]: Is it really only the compression function and not the full hash here?

File: stdlib/hashes/sha256/IVconstants.code [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
...3[4], h3[5], h3[6], h3[7], h3[8], h3[9], h3[10], h3[11], h3[12], h3[13], h3[14], h3[15], h3[16], h3[17], h3[18], h3[19], h3[20], h3[21], h3[22], h3[23], h3[24], h3[25], h3[26], h3[27], h3[28], h3[29], h3[30], h3[31], h4[0], h4[1], h4[2], h4[3], h4[4], h4[5], h4[6], h4[7], h4[8], h4[9], h4[10], h4[11], h4[12], h4[13], h4[14], h4[15], h4[16], h4[17], h4[18], h4[19], h4[20], h4[21], h4[22], h4[23], h4[24], h4[25], h4[26], h4[27], h4[28], h4[29], h4[30], h4[31], h5[0], h5[1], h5[2], h5[3], h5[4], h5[5], h5[6], h5[7], h5[8], h5[9], h5[10], h5[11], h5[12], h5[13], h5[14], h5[15], h5[16], h5[17], h5[18], h5[19], h5[20], h5[21], h5[22], h5[23], h5[24], h5[25], h5[26], h5[27], h5[28], h5[29], h5[30], h5[31], h6[0], h6[1], h6[2], h6[3], h6[4], h6[5], h6[6], h6[7], h6[8], h6[9], h6[10], h6[11], h6[12], h6[13], h6[14], h6[15], h6[16], h6[17], h6[18], h6[19], h6[20], h6[21], h6[22], h6[23], h6[24], h6[25], h6[26], h6[27], h6[28], h6[29], h6[30], h6[31], h7[0], h7[1], h7[2], h7[3], h7[4], h7[5], h7[6], h7[7], h7[8], h7[9], h7[10], h7[11], h7[12], h7[13], h7[14], h7[15], h7[16], h7[17], h7[18], h7[19], h7[20], h7[21], h7[22], h7[23], h7[24], h7[25], h7[26], h7[27], h7[28], h7[29], h7[30], h7[31]]
```
Comment (Schaeff) [inline stdlib/hashes/sha256/IVconstants.code]: Is it worth opening an issue on how to make this operation less verbose in the DSL?

File: stdlib/hashes/sha256/shaRound.code [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import "LIBSNARK/sha256round" as sha256
+
+def main(field[256] a, field[256] b, field[256] IV) -> (field[256]):
+
+  // a and b is NOT checked to be of type bool
```
Comment (Schaeff) [inline stdlib/hashes/sha256/shaRound.code]: maybe we could call such files `shaRoundUnchecked` to signal they're not binary checked?

File: stdlib/merkleTree/sha256PathProof.code [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import "../hashes/sha256/512bit.code" as sha256
+import "../utils/multiplexer/256bit.code" as multiplex
+import "../utils/binary/not.code" as NOT
+
+// Merke-Tree inclusion proof for tree depth 3
```
Comment (Schaeff) [inline stdlib/merkleTree/sha256PathProof.code]: As far as I understand, we can't make this generic over `treeDepth` yet so maybe the file should be named accordingly (for depth 3 here) `sha256PathProof3` or something

File: stdlib/merkleTree/sha256PathProof.code [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+	lhs = multiplex(currentDirection, currentDigest, PathDigest1)
+	rhs = multiplex(NOT(currentDirection), currentDigest, PathDigest1)
+	currentDigest = sha256(lhs, rhs)
+	counter = counter + 1
+
+	//Asserts
+	counter == treeDepth
+	rootDigest == currentDigest
+
+	return 1 //return true in success
```
Comment (Schaeff) [inline stdlib/merkleTree/sha256PathProof.code]: I can't comment on them directly as they were only moved, but the binary adder is extremely inefficient compared to a nondeterministic approach, any reason to have them in stdlib?

File: stdlib/utils/multiplexer/256bit.code [hunk 6]
```diff
@@ -0,0 +1,11 @@
+def main(field selector, field[256] lhs, field[256] rhs) -> (field[256]):
+
+	field[256] out = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+
+	for field i in 0..256 do
+
```
Comment (Schaeff) [inline stdlib/utils/multiplexer/256bit.code]: Why extra \n ?

File: stdlib/utils/multiplexer/2bit.code [hunk 7]
```diff
@@ -0,0 +1,11 @@
+def main(field selector, field[2] lhs, field[2] rhs) -> (field[2]):
+
+	field[2] out = [0, 0] 
+
+	for field i in 0..2 do
+
```
Comment (Schaeff) [inline stdlib/utils/multiplexer/2bit.code]: Same \n

File: zokrates_book/src/concepts/imports.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+```
+
+There also is a handy syntax to import from the parent directory:
+```zokrates
+import "../mycode.code"
+```
+
+Also imports further up the file-system are supported:
+```zokrates
+import "./../../../mycode.code"
```
Comment (Schaeff) [inline zokrates_book/src/concepts/imports.md]: `../` should work directly here

File: zokrates_book/src/concepts/imports.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+```
+
+You can also choose to rename the imported resource, like so:
+```zokrates
+import "./mycode.code" as abc
+```
+
+### Absolute Imports
+
+Absolute imports don't start with `./` in the path and are used to import components from the ZoKrates standard library. Please check the according [section](./stdlib.html) for more details.
```
Comment (Schaeff) [inline zokrates_book/src/concepts/imports.md]: `../` is also relative

File: zokrates_book/src/concepts/stdlib.md [hunk 10]
```diff
@@ -1,40 +1,49 @@
 ## Standard library
 
-ZoKrates comes with a number of reusable components. For now, these components are:
+ZoKrates comes with a number of reusable components which are defined at `./stdlib/` in the ZoKrates root folder. The standard library is solely based on the ZoKrates DSL and can be easily extended.
```
Comment (Schaeff) [inline zokrates_book/src/concepts/stdlib.md]: A user would not have access to the source code, so should it rather be `$ZOKRATES_HOME` for them?

File: zokrates_book/src/concepts/stdlib.md [hunk 11]
```diff
@@ -1,40 +1,49 @@
...
-import "LIBSNARK/sha256packed"
+import "hashes/sha256/512bitPacked.code" 
 ```
 
-A function that takes 4 field elements as inputs, unpacks each of them to 128 bits (big endian), concatenates them and applies sha256. It then returns two field elements, each representing 128 bits of the result.
+A function that takes an array of 4 field elements as inputs, unpacks each of them to 128 bits (big endian), concatenates them and applies sha256. It then returns an array of 2 field elements, each representing 128 bits of the result.
+
+### Direct imports
+
+Some components of the standard library can not be efficiently represented in the ZoKrates DSL language. Those functions are injected at compile-time and are available by default.
```
Comment (Schaeff) [inline zokrates_book/src/concepts/stdlib.md]: *cannot

Comment (Schaeff) [reply zokrates_book/src/concepts/stdlib.md]: *+yet

File: zokrates_book/src/sha256example.md [hunk 12]
```diff
@@ -13,7 +13,7 @@ Make sure you have followed the instructions in the [Getting Started](gettingsta
 
 We will start this tutorial by using ZoKrates to compute the hash for an arbitrarily chosen preimage, being the number `5` in this example.
 
-First, we create a new file named `hashexample.code` with the following content:
+First, make sure that the `zokrates` binary is located in the `ZoKrates` directory. Then, we create a new file named `hashexample.code` with the following content:
```
Comment (Schaeff) [inline zokrates_book/src/sha256example.md]: Why is that? If it's needed, do we have an easy way to make that more seamless?

File: zokrates_cli/Cargo.toml [hunk 13]
```diff
@@ -6,6 +6,7 @@ repository = "https://github.com/JacobEberhardt/ZoKrates.git"
 edition = "2018"
 
 [features]
+# default = []
```
Comment (Schaeff) [inline zokrates_cli/Cargo.toml]: to be removed?

File: zokrates_cli/examples/TestStdlib/hashes/sha256/test512bitPacked.code [hunk 14]
```diff
@@ -0,0 +1,14 @@
+import "hashes/sha256/512bitPacked.code" as sha256packed
```
Comment (Schaeff) [inline zokrates_cli/examples/TestStdlib/hashes/sha256/test512bitPacked.code]: would be good to include the python snippet of how the test case was generated?

File: zokrates_core/lib/wraplibsnarkgadgets.hpp [hunk 15]
```diff
@@ -12,11 +12,13 @@ extern "C" {
...
 
-char* _sha256Constraints();
-char* _sha256Witness(const uint8_t* input, int input_length);
+char* _sha256RoundConstraints();
+char* _sha256RoundWitness(const uint8_t* input, int input_length);
 
-char* _shaEth256Constraints();
-char* _shaEth256Witness(const uint8_t* input, int input_length);
+// External interface to free memory 
+void _free_string(char const *str) {
```
Comment (Schaeff) [inline zokrates_core/lib/wraplibsnarkgadgets.hpp]: THis is fine for now but if this is not as clean as we'd want, let's open an issue.

File: .circleci/config.yml [hunk 16]
```diff
@@ -14,7 +14,7 @@ jobs:
           command: cargo generate-lockfile
       - restore_cache:
           keys:
-            - v4-cargo-cache-{{ arch }}-{{ checksum "Cargo.lock" }}
+            - v5-cargo-cache-{{ arch }}-{{ checksum "Cargo.lock" }}
```
[the remark this answers is not in the document]
Comment (stefandeml) [reply .circleci/config.yml]: sure.

File: stdlib/hashes/sha256/IVconstants.code (hunk 2 above)
Comment (stefandeml) [reply stdlib/hashes/sha256/IVconstants.code]: Good idea. Here you go: https://github.com/Zokrates/ZoKrates/issues/246

File: stdlib/hashes/sha256/shaRound.code (hunk 3 above)
Comment (stefandeml) [reply stdlib/hashes/sha256/shaRound.code]: Good idea. Change to `shaRoundNoBoolCheck.code`

File: zokrates_book/src/sha256example.md (hunk 12 above)
Comment (stefandeml) [reply zokrates_book/src/sha256example.md]: not needed any more. this was still based on relative imports when we didn't have support for absolute imports

File: stdlib/merkleTree/sha256PathProof.code (hunk 5 above)
Comment (stefandeml) [reply stdlib/merkleTree/sha256PathProof.code]: moved all sha256 related tools back to the example folder

File: .circleci/config.yml [hunk 17]
```diff
@@ -14,13 +14,13 @@ jobs:
...
```
Comment (Schaeff) [inline .circleci/config.yml]: can be reverted

File: stdlib/hashes/sha256/512bitPacked.code [hunk 18]
```diff
@@ -0,0 +1,23 @@
+import "PACKING/pack128" as pack128
+import "PACKING/unpack128" as unpack128
+import "./512bit.code" as sha256
+//FIXME: delete packed stldib
```
[the remark this answers is not in the document]
Comment (stefandeml) [reply stdlib/hashes/sha256/512bitPacked.code]: yes, sha256packed is not part of `imports.rs` any more

File: zokrates_cli/examples/TestStdlib/hashes/sha256/test512bitPadded.code [hunk 19]
```diff
@@ -0,0 +1,11 @@
+import "hashes/sha256/512bitPadded.code" as sha256
```
[the remark this answers is not in the document]
Comment (stefandeml) [reply zokrates_cli/examples/TestStdlib/hashes/sha256/test512bitPadded.code]: Added.

File: zokrates_cli/examples/TestStdlib/merkleTree/testsha256PathProof3.code [hunk 20]
```diff
@@ -0,0 +1,13 @@
+import "merkleTree/sha256PathProof3.code" as merkleTreeProof
```
[the remark this answers is not in the document]
Comment (stefandeml) [reply zokrates_cli/examples/TestStdlib/merkleTree/testsha256PathProof3.code]: I'm not aware of a sha256 compressed implementation in order to create an example with Python.

File: .circleci/config.yml (hunk 17 above)
Comment (stefandeml) [reply .circleci/config.yml]: Sorry, fixed.

File: zokrates_cli/examples/TestStdlib/merkleTree/testsha256PathProof3.code (hunk 20 above)
[the remark this answers is not in the document]
Comment (stefandeml) [reply zokrates_cli/examples/TestStdlib/merkleTree/testsha256PathProof3.code]: Moved gadget to examples folder

**Pull request merged**